### PR TITLE
[3545] Page components should support drag-drop of embedded components

### PR DESCRIPTION
**Pull request by aachavarria**
https://github.com/craftercms/craftercms/issues/3545

### CODE REVIEW

File: static-assets/components/cstudio-common/resources/en/base.js [hunk 1]
```diff
@@ -537,6 +537,7 @@ CStudioAuthoring.Messages.registerBundle("forms", "en", {
     unableLoad: "The system was unable to load ",
     failure: "Failure.",
     loadFormDefError: "Failed to load form definition.",
+    componentsNotSupported: "The drop zone not support this type of Components",
```
Comment (rart) [inline static-assets/components/cstudio-common/resources/en/base.js:540]: Use new mechanism or include in all bundles.

File: static-assets/components/cstudio-forms/controls/node-selector.js [hunk 2]
```diff
@@ -504,16 +511,19 @@ YAHOO.extend(CStudioForms.Controls.NodeSelector, CStudioForms.CStudioFormField,
             }
 
             item.datasource = datasource;
-            this.items[this.items.length] = item
+            this.items[this.items.length] = item;
 
-            if(this.form.datasourceMap[datasource].itemsAreContentReferences) {
+            if(this.form.datasourceMap[datasource] && this.form.datasourceMap[datasource].itemsAreContentReferences) {
```
Comment (rart) [inline static-assets/components/cstudio-forms/controls/node-selector.js]: This feels dangerous for preexisting situations covered

File: static-assets/components/cstudio-forms/forms-engine.js [hunk 3]
```diff
@@ -838,12 +838,22 @@ var CStudioForms = CStudioForms || function() {
...
+                  value:message.value,
+                  inline:true, selectorId:
+                  message.selectorId,
+                  ds: message.ds
+                });
+              } else{
+                amplify.publish('UPDATE_NODE_SELECTOR', message );
+              }
+
+              //cfe.engine.saveForm(false, message.draft, true);
```
Comment (rart) [inline static-assets/components/cstudio-forms/forms-engine.js]: Remove if unused

File: static-assets/components/cstudio-preview-tools/mods/component-panel.js [hunk 4]
```diff
@@ -193,67 +193,182 @@
...
+                      }
+                    } else {
+                      var CMgs = CStudioAuthoring.Messages;
+                      var langBundle = CMgs.getBundle("forms", CStudioAuthoringContext.lang);
+                      CStudioAuthoring.Operations.showSimpleDialog(
+                        "componentsNotSupported-dialog",
+                        CStudioAuthoring.Operations.simpleDialogTypeINFO,
+                        CMgs.format(langBundle, "notification"),
+                        CMgs.format(langBundle, "componentsNotSupported"),
+                        [{ text: "OK",  handler:function(){
```
Comment (rart) [inline static-assets/components/cstudio-preview-tools/mods/component-panel.js]: reformat

File: static-assets/components/cstudio-preview-tools/mods/component-panel.js [hunk 5]
```diff
@@ -193,67 +193,182 @@
...
+
+                validate: function(componentPath, pageContentType, zones, componentType) {
+                  var zone = Object.keys(zones)[0];
 
-                    if (isNew == true) {
-                      var subscribeCallback = function (_message) {
-                        switch (_message.type) {
-                          case "FORM_CANCEL": {
-                            amplify.unsubscribe('FORM_ENGINE_MESSAGE_POSTED', subscribeCallback);
+                  return new Promise((resolve, reject) => {
```
Comment (rart) [inline static-assets/components/cstudio-preview-tools/mods/component-panel.js:303]: cache as you go and resolve immediately if already in cache

File: static-assets/components/cstudio-preview-tools/mods/component-panel.js [hunk 6]
```diff
@@ -193,67 +193,182 @@
...
+                        });
+                        break;
+                      }
+                    }
+                  };
+                  amplify.subscribe('FORM_ENGINE_MESSAGE_POSTED', subscribeCallback);
+                  var parentPath = compPath || CStudioAuthoring.ComponentsPanel.getPreviewPagePath(
+                    CStudioAuthoringContext.previewCurrentPath);
+
+                  var getContentItemsCb = {
```
Comment (rart) [inline static-assets/components/cstudio-preview-tools/mods/component-panel.js]: put callback directly on the call

File: static-assets/scripts/dnd-controller.js [hunk 7]
```diff
@@ -389,6 +389,7 @@ define('dnd-controller', ['crafter', 'jquery', 'jquery-ui', 'animator', 'communi
     var compPath = $dropZone.parents('[data-studio-component-path]').attr('data-studio-component-path');
     var compTracking = $dropZone.parents('[data-studio-component-path]').attr('data-studio-tracking-number');
     var objectId = $dropZone.attr('data-studio-components-objectid');
+    var embeddedId = $dropZone.attr('data-studio-embedded-item-id');
```
Comment (rart) [inline static-assets/scripts/dnd-controller.js]: what would this be? where would it come from?

File: static-assets/components/cstudio-forms/controls/node-selector.js (hunk 2 above)
Comment (aachavarria) [reply static-assets/components/cstudio-forms/controls/node-selector.js]: will removed, this is not needed

File: static-assets/scripts/dnd-controller.js (hunk 7 above)
Comment (aachavarria) [reply static-assets/scripts/dnd-controller.js]: still pending to use, will check if need it to move elements

**Review by rart — CHANGES_REQUESTED** [review]
Revert App.tsx

**Review by rart — CHANGES_REQUESTED** [review]
Revert App.tsx

**Review by rart — APPROVED** [review]

**Pull request merged**